options/configurable.cc

```diff
@@ -0,0 +1,606 @@
+  int found = 1;
+  std::unordered_set<std::string> unsupported;
+  // While there are unused properties and we processed at least one,
+  // go through the remaining unused properties and attempt to configure them.
+  while (found > 0 && !options->empty()) {
+    found = 0;
+    notsup = Status::OK();
+    for (auto it = options->begin(); it != options->end();) {
+      const std::string& opt_name = configurable.GetOptionName(it->first);
+      const std::string& opt_value = it->second;
+      const auto opt_info =
+          OptionTypeInfo::Find(opt_name, type_map, &elem_name);
+      if (opt_info == nullptr) {  // Did not find the option.  Skip it
+        ++it;
+      } else {
+        Status s = ConfigureOption(config_options, configurable, *opt_info,
+                                   opt_name, elem_name, opt_value, opt_ptr);
+        if (s.IsNotFound()) {
+          ++it;
+        } else if (s.IsNotSupported()) {
+          notsup = s;
+          unsupported.insert(it->first);
+          ++it;  // Skip it for now
+        } else {
+          found++;
+          it = options->erase(it);
+          if (!s.ok()) {
+            result = s;
+          }
+        }
+      }
+    }  // End for all remaining options
+  }    // End while found one or options remain
+
+  // Now that we have been through the list, remove any unsupported
+  for (auto u : unsupported) {
+    auto it = options->find(u);
+    if (it != options->end()) {
+      options->erase(it);
+    }
+  }
+  if (config_options.ignore_unknown_options) {
+    if (!result.ok()) result.PermitUncheckedError();
+    if (!notsup.ok()) notsup.PermitUncheckedError();
+    return Status::OK();
+  } else if (!result.ok()) {
+    if (!notsup.ok()) notsup.PermitUncheckedError();
+    return result;
+  } else if (config_options.ignore_unsupported_options) {
+    if (!notsup.ok()) notsup.PermitUncheckedError();
+    return Status::OK();
+  } else {
+    return notsup;
+  }
+}
+
+Status ConfigurableHelper::ConfigureSingleOption(
+    const ConfigOptions& config_options, Configurable& configurable,
+    const std::string& name, const std::string& value) {
+  std::string opt_name;
+  void* opt_ptr = nullptr;
+  const auto opt_info =
+      FindOption(configurable.options_, name, &opt_name, &opt_ptr);
+  if (opt_info == nullptr) {
+    return Status::NotFound("Could not find option: ", name);
+  } else {
+    return ConfigureOption(config_options, configurable, *opt_info, name,
+                           opt_name, value, opt_ptr);
+  }
+}
+
+Status ConfigurableHelper::ConfigureOption(
+    const ConfigOptions& config_options, Configurable& configurable,
+    const OptionTypeInfo& opt_info, const std::string& opt_name,
+    const std::string& name, const std::string& value, void* opt_ptr) {
+  if (opt_name == name) {
+    return configurable.ParseOption(config_options, opt_info, opt_name, value,
+                                    opt_ptr);
+  } else if (opt_info.IsStruct()) {
```

Comment:
We can combine the condition as `if (opt_info.IsStruct() || opt_info.IsConfigurable())`

Reply:
Done